### DISCUSSION TITLE
Understanding the Message Passing class

### DISCUSSION THREAD

**Discussion by mdanb** [pyg-team/pytorch_geometric#2378] [Q&A] [answered]
I'm trying to gain a better understanding of the Message Passing class so I'm looking at the source code in `torch_geometric/nn/conv/message_passing.py`. I'm struggling to understand why the `__init__` does what it does. In particular:
1- What do the `special_args` represent semantically?
2- In:
```
self.inspector.inspect(self.message)
self.inspector.inspect(self.aggregate, pop_first=True)
self.inspector.inspect(self.message_and_aggregate, pop_first=True)
self.inspector.inspect(self.update, pop_first=True)
```
why `pop_first` for everything except `self.message`, and why do the popping in the first place?

**Comment by rusty1s** [comment, marked as answer]
1. `special_args` denote the arguments that are automatically filled and which can be additionally accessed in `message`, `aggregate`, `update`.
2. `aggregate` and `update` are defined to take the output of the previous call as first argument for input. Therefore, we drop these arguments from the argument list.

**Reply by mdanb** [reply]
Can you elaborate further? 
1. I'm not quite sure what you mean by "automatically filled" and "can additionally access"
2. What do you mean by "previous call"?

**Reply by rusty1s** [reply]
Sure, sorry for being so fuzzy :)

1. The `special_args` holds all additional argument names one can access in `message`, `aggregate`, and `update`, e.g., you can access the `index` along which to aggregate, or the `size_i` of the output number of nodes (basically some additional information about the graph). For example, this allows you to do the following:
```python
    self.propagate(edge_index, x=x)

def message(self, x_i, x_j, index, size_i):
    ...
```
2. The message passing flow we follow is: `message -> aggregate -> update`. With this, `aggregate` will receive the output of `message` as its first argument. The same holds true for `update`.
